proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,493 @@
+the actual lifetime of `y` to release it after its last use, there isn't
+a strong guarantee that it will. Even if the optimizer does what we want,
+programmers modifying this code in the future
+may introduce new references to `y` that inadvertently extend its lifetime
+and break our attempt to keep `x` unique. There isn't any indication in the
+source code that that the end of `y`'s use is important to the performance
+characteristics of the code.
+
+Swift-evolution pitch thread: [https://forums.swift.org/t/pitch-move-function-use-after-move-diagnostic](https://forums.swift.org/t/pitch-move-function-use-after-move-diagnostic)
+
+## Proposed solution: Move Function + "Use After Move" Diagnostic
+
+That is where the `move` function comes into play. The `move` function consumes
+a **movable binding**, which is either
+an unescaped local `let`, unescaped local `var`, or function argument, with
+no property wrappers or get/set/read/modify/etc. accessors applied. It then
+ provides a compiler guarantee that the binding will
+be unable to be used again locally. If such a use occurs, the compiler will
+emit an error diagnostic. We can modify the previous example to use `move` to
+explicitly end the lifetime of `y` when we're done with it:
+
+```swift
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  // We no longer use y after this point, so move it when we pass it off to
+  // the last use.
+  consumeFinalY(move(y))
+
+  // x will be unique again here.
+  x.append(7)
+}
+```
+
+This addresses both of the motivating issues above: `move` guarantees the
+lifetime of `y` ends at the given point, allowing the compiler to generate
+code to clean up or transfer ownership of `y` without relying on optimization.
+Furthermore, if a future maintainer modifies the code in a way that extends
+the lifetime of `y` past the expected point, then the compiler will raise an
+error. For instance, if we try:
+
+```swift
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  // We think we no longer use y after this point...
+  consumeFinalY(move(y))
+
+  // ...and x will be unique again here...
+  x.append(7)
+
+  // ...but this additional use of y snuck in:
+  useYAgain(y)
+}
+```
+
+In this case, we get the following output from the compiler as expected:
+
+```swift
+test.swift:10:7: error: 'y' used after being moved
```

Comment:
I would drop this entire example and just put the error message directly into the example above:
```
   useYAgain(y) // Error: 'y' used after being moved
```

Reply:
Done.